upgrades/upgradevalidation/upgrade.go

```diff
@@ -9,62 +9,53 @@ import (
 	environscloudspec "github.com/juju/juju/environs/cloudspec"
 )
 
-// ValidatorsForControllerUpgrade returns a list of validators for controller upgrade,
+// ValidatorsForControllerUpgrade returns a list of validators for controller
+// upgrade.
+// Note: the target version can never be lower than the current version.
 func ValidatorsForControllerUpgrade(
 	isControllerModel bool, targetVersion version.Number, cloudspec environscloudspec.CloudSpec,
 ) []Validator {
 	if isControllerModel {
 		validators := []Validator{
 			getCheckTargetVersionForControllerModel(targetVersion),
 			checkMongoStatusForControllerUpgrade,
+			checkMongoVersionForControllerModel,
+			checkNoWinMachinesForModel,
+			checkForDeprecatedUbuntuSeriesForModel,
+			getCheckForLXDVersion(cloudspec),
 		}
-		if targetVersion.Major == 3 {
-			validators = append(validators,
-				checkMongoVersionForControllerModel,
-				checkNoWinMachinesForModel,
-				checkForDeprecatedUbuntuSeriesForModel,
-				getCheckForLXDVersion(cloudspec),
-			)
-			if targetVersion.Minor >= 1 {
-				validators = append(validators, checkForCharmStoreCharms)
-			}
+		if targetVersion.Major == 3 && targetVersion.Minor >= 1 {
+			validators = append(validators, checkForCharmStoreCharms)
 		}
 		return validators
 	}
+
 	validators := []Validator{
 		getCheckTargetVersionForModel(targetVersion, UpgradeControllerAllowed),
 		checkModelMigrationModeForControllerUpgrade,
+		checkNoWinMachinesForModel,
+		checkForDeprecatedUbuntuSeriesForModel,
+		getCheckForLXDVersion(cloudspec),
 	}
-	if targetVersion.Major >= 3 {
-		validators = append(validators,
-			checkNoWinMachinesForModel,
-			checkForDeprecatedUbuntuSeriesForModel,
-			getCheckForLXDVersion(cloudspec),
-		)
-	}
-	if targetVersion.Major >= 3 && targetVersion.Minor >= 1 {
+	if targetVersion.Major == 3 && targetVersion.Minor >= 1 {
 		validators = append(validators, checkForCharmStoreCharms)
 	}
-
 	return validators
 }
 
-// ValidatorsForModelUpgrade returns a list of validators for model upgrade,
+// ValidatorsForModelUpgrade returns a list of validators for model upgrade.
+// Note: the target version can never be lower than the current version.
 func ValidatorsForModelUpgrade(
 	force bool, targetVersion version.Number, cloudspec environscloudspec.CloudSpec,
 ) []Validator {
 	validators := []Validator{
 		getCheckUpgradeSeriesLockForModel(force),
+		checkNoWinMachinesForModel,
+		checkForDeprecatedUbuntuSeriesForModel,
+		getCheckForLXDVersion(cloudspec),
 	}
-	if targetVersion.Major == 3 {
-		validators = append(validators,
-			checkNoWinMachinesForModel,
-			checkForDeprecatedUbuntuSeriesForModel,
-			getCheckForLXDVersion(cloudspec),
-		)
-		if targetVersion.Minor >= 1 {
-			validators = append(validators, checkForCharmStoreCharms)
-		}
+	if targetVersion.Major == 3 && targetVersion.Minor >= 1 {
+		validators = append(validators, checkForCharmStoreCharms)
```

Comment:
I'm not hitting this check. Migration doesn't fail until here:
```machine-0: 20:44:19 ERROR juju.worker.migrationmaster.a98b55 model data transfer failed, failed to migrate binaries: cannot upload charms: cannot upload charm: Post https://10.186.60.233:17070/migrate/charms?arch=&name=ubuntu&revision=24&schema=cs&series=&user=: cannot upload charm: unsupported schema "cs"```

Reply:
Yeah because you're not allowed cs charms in 3.1. So this is correct.

Reply:
So I did some investigation. This is entirely correct for what we have now. If you try and migrate from 2.9 to 3.1 there is no check at all, on the source side.

https://github.com/juju/juju/blob/02d498631e196f2a37f9b7c3b5c31bdcb1dad333/upgrades/upgradevalidation/migrate.go#L13-L28

The target side will only ever validate from 3.1 onwards

https://github.com/juju/juju/blob/d3c8991a8d1be1f5711ae53e096caadeba3244a3/upgrades/upgradevalidation/upgrade.go#L21-L31

So this PR doesn't actually break any existing flow from 2.9 to 3.1. It's currently been this way. There is no way to get a cs charm into 3.1, so we could actually remove the check on the source side as well.

Here are my steps to reproduce exactly the same thing as this comment:

```sh
$ snap install juju --classic --channel=2.9/stable
$ /snap/bin/juju bootstrap lxd src29
$ /snap/bin/juju deploy cs:~jameinel/ubuntu-lite
```

```sh
$ git checkout 3.1
$ juju bootstrap lxd target31 --build-agent
```

```sh
$ /snap/bin/juju migrate src29:default target31
```

@hmlanigan we can fix that issue in a subsequent PR, but not this one, then I'm fine with that.